README.md
# status_stepper

Use this package for showing status changing. For now available only left to right animation.

![](https://github.com/Progressive-Mobile/status_stepper/blob/0ed01e34803e66da47a8aabc68a0ee6367e3cc53/readme_resources/status_stepper.gif)

## Quick reference

Property | What does it do
-------- | ---------------
animationDuration    | The duration of animation for 1 stepper item, multiplyes by 2(for animating connector and item) * n(n - count of children to animate)
animationDelayDuration    | Duration before starting animation 
currentIndex    | Current status index, -1 if there are no  active items
lastActiveIndex    | Animation starts after widget at this position 
activeColor    | Color of active and passed statuses, by default is Theme.of(context).primaryColor 
disabledColor    | Color of next statuses, by default is Theme.of(context).colorScheme.secondaryVariant 
connectorCurve    | Curve for the connectors
itemCurve    | Curve for the status widgets
connectorThickness    | Thickness of the connector 

## Example

```dart
class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Status Stepper Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Status Stepper Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final statuses = List.generate(
    4,
    (index) => SizedBox.square(
      dimension: 32,
      child: Center(child: Text('$index')),
    ),
  );

  int curIndex = -1;
  int lastIndex = -1;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: StatusStepper(
                connectorCurve: Curves.easeIn,
                itemCurve: Curves.easeOut,
                activeColor: Colors.amber,
                disabledColor: Colors.grey,
                animationDuration: const Duration(milliseconds: 500),
                children: statuses,
                lastActiveIndex: lastIndex,
                currentIndex: curIndex,
                connectorThickness: 6,
              ),
            ),
            ElevatedButton(
              onPressed: () {
                setState(() {
                  curIndex = -1;
                  lastIndex = -1;
                });
              },
              child: const Text(
                'Reset',
              ),
            ),
            Expanded(
              child: ListView.builder(
                itemCount: statuses.length,
                itemBuilder: (context, index) => ElevatedButton(
                  onPressed: index > curIndex
                      ? () {
                          setState(() {
                            lastIndex = curIndex;
                            curIndex = index;
                          });
                        }
                      : null,
                  child: Text(
                    '$index',
                  ),
                ),
              ),
            )
          ],
        ),
      ),
    );
  }
}
```

CHANGELOG.md
## [1.0.0] - 2021-01-11

### Added

- Status stepper creation
- Status stepper decoration 

lib/stepper_item.dart
import 'package:flutter/material.dart';

class StepperItem extends StatelessWidget {
  final Widget child;
  final bool isPassed;
  final bool shouldRedraw;
  final int delayFactor;
  final Duration animationDuration;
  final Duration animationAwaitDuration;
  final Color? activeColor;
  final Color? disabledColor;
  final Curve curve;

  const StepperItem({
    Key? key,
    required this.isPassed,
    required this.child,
    required this.animationDuration,
    required this.shouldRedraw,
    required this.delayFactor,
    required this.curve,
    this.animationAwaitDuration = Duration.zero,
    this.activeColor,
    this.disabledColor,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ClipOval(
      child: Container(
        color: disabledColor ?? Theme.of(context).colorScheme.secondaryVariant,
        child: Stack(
          children: [
            FutureBuilder(
              future: Future.delayed(
                animationDuration * delayFactor + animationAwaitDuration,
              ),
              builder: (context, snapshot) => AnimatedSwitcher(
                transitionBuilder: (child, animation) => SizeTransition(
                  sizeFactor: animation,
                  child: child,
                  axis: Axis.horizontal,
                ),
                switchInCurve: curve,
                duration: animationDuration,
                child: isPassed && snapshot.connectionState == ConnectionState.done || !shouldRedraw
                    ? Container(
                        color: activeColor ?? Theme.of(context).primaryColor,
                        alignment: Alignment.centerLeft,
                        child: child,
                        foregroundDecoration: BoxDecoration(
                          color: activeColor ?? Theme.of(context).primaryColor,
                        ),
                      )
                    : const SizedBox(),
              ),
            ),
            child,
          ],
        ),
      ),
    );
  }
}


lib/status_stepper.dart
library progress_stepper;

import 'package:flutter/material.dart';

import 'stepper_connector.dart';
import 'stepper_item.dart';

class StatusStepper extends StatelessWidget {
  ///Duration multiplyes by 2(for animating connector and item) * n(n - count of children to animate)
  final Duration animationDuration;

  ///Duration before starting animation
  final Duration animationDelayDuration;

  ///Current status index, -1 if there are no  active items
  final int currentIndex;

  ///Animation starts after widget at this position
  final int lastActiveIndex;

  final List<Widget> children;

  ///Color of active and passed statuses, by default is Theme.of(context).primaryColor
  final Color? activeColor;

  ///Color of next statuses, by default is Theme.of(context).colorScheme.secondaryVariant
  final Color? disabledColor;

  ///Curve for the connectors
  final Curve connectorCurve;

  ///Curve for the status widgets
  final Curve itemCurve;

  final double connectorThickness;

  const StatusStepper({
    required this.children,
    this.animationDuration = const Duration(milliseconds: 200),
    this.lastActiveIndex = -1,
    this.currentIndex = -1,
    this.animationDelayDuration = Duration.zero,
    this.connectorCurve = Curves.linear,
    this.itemCurve = Curves.linear,
    this.connectorThickness = 2,
    this.activeColor,
    this.disabledColor,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final _children = <Widget>[];
    for (var index = 0; index < children.length; index++) {
      final delayFactor = index - lastActiveIndex > 1 ? index - lastActiveIndex - 1 : 0;
      final isNotCurrentIndex = currentIndex != lastActiveIndex;
      _children.addAll(
        [
          if (index != 0)
            StepperConnector(
              animationDuration: animationDuration,
              isPassed: index <= currentIndex,
              shouldRedraw: index > lastActiveIndex,
              delayFactor: delayFactor * 2 - (!isNotCurrentIndex && delayFactor > 0 ? 1 : 0),
              animationAwaitDuration: animationDelayDuration,
              activeColor: activeColor,
              disabledColor: disabledColor,
              curve: connectorCurve,
              connectorThickness: connectorThickness,
            ),
          StepperItem(
            animationDuration: animationDuration,
            child: children[index],
            isPassed: index <= currentIndex,
            shouldRedraw: index > lastActiveIndex,
            delayFactor: delayFactor * 2 + (isNotCurrentIndex ? 1 : 0),
            animationAwaitDuration: animationDelayDuration,
            activeColor: activeColor,
            disabledColor: disabledColor,
            curve: itemCurve,
          ),
        ],
      );
    }

    return Row(
      children: _children,
    );
  }
}


lib/stepper_connector.dart
import 'package:flutter/material.dart';

class StepperConnector extends StatelessWidget {
  final bool isPassed;
  final bool shouldRedraw;
  final int delayFactor;
  final Duration animationDuration;
  final Duration animationAwaitDuration;
  final Color? activeColor;
  final Color? disabledColor;
  final Curve curve;
  final double connectorThickness;

  const StepperConnector({
    Key? key,
    required this.isPassed,
    required this.delayFactor,
    required this.shouldRedraw,
    required this.animationDuration,
    required this.curve,
    required this.connectorThickness,
    this.animationAwaitDuration = Duration.zero,
    this.activeColor,
    this.disabledColor,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Expanded(
      child: Stack(
        children: [
          Divider(
            thickness: connectorThickness,
            color: disabledColor ?? Theme.of(context).colorScheme.secondaryVariant,
          ),
          FutureBuilder(
            future: Future.delayed(
              animationDuration * delayFactor + animationAwaitDuration,
            ),
            builder: (context, snapshot) => AnimatedSwitcher(
              switchInCurve: curve,
              transitionBuilder: (child, animation) => SizeTransition(
                sizeFactor: animation,
                child: child,
                axis: Axis.horizontal,
              ),
              duration: animationDuration,
              child: isPassed && snapshot.connectionState == ConnectionState.done || !shouldRedraw
                  ? Divider(
                      thickness: connectorThickness,
                      color: activeColor ?? Theme.of(context).primaryColor,
                    )
                  : const SizedBox(),
            ),
          ),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:status_stepper/status_stepper.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Status Stepper Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Status Stepper Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final statuses = List.generate(
    4,
    (index) => SizedBox.square(
      dimension: 32,
      child: Center(child: Text('$index')),
    ),
  );

  int curIndex = -1;
  int lastIndex = -1;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: StatusStepper(
                connectorCurve: Curves.easeIn,
                itemCurve: Curves.easeOut,
                activeColor: Colors.amber,
                disabledColor: Colors.grey,
                animationDuration: const Duration(milliseconds: 500),
                children: statuses,
                lastActiveIndex: lastIndex,
                currentIndex: curIndex,
                connectorThickness: 6,
              ),
            ),
            ElevatedButton(
              onPressed: () {
                setState(() {
                  curIndex = -1;
                  lastIndex = -1;
                });
              },
              child: const Text(
                'Reset',
              ),
            ),
            Expanded(
              child: ListView.builder(
                itemCount: statuses.length,
                itemBuilder: (context, index) => ElevatedButton(
                  onPressed: index > curIndex
                      ? () {
                          setState(() {
                            lastIndex = curIndex;
                            curIndex = index;
                          });
                        }
                      : null,
                  child: Text(
                    '$index',
                  ),
                ),
              ),
            )
          ],
        ),
      ),
    );
  }
}
